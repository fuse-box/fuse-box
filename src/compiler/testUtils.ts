import { ICompilerOptions } from '../compilerOptions/interfaces';
import { transformModule } from './core/transformModule';
import { generate } from './generator/generator';
import { ASTNode } from './interfaces/AST';
import { ITransformer, ITransformerCommon } from './interfaces/ITransformer';
import { ITransformerRequireStatementCollection } from './interfaces/ITransformerRequireStatements';
import { ImportType } from './interfaces/ImportType';
import { parseTypeScript } from './parser';
import { ISerializableTransformationContext } from './transformer';

function cleanupForTest(node) {
  delete node.loc;
  delete node.raw;
  delete node.range;
}
export function initCommonTransform(props: {
  code: string;
  compilerOptions?: ICompilerOptions;
  jsx?: boolean;
  props?: ISerializableTransformationContext;
  transformers: Array<ITransformer>;
}) {
  const requireStatementCollection: ITransformerRequireStatementCollection = [];
  function onRequireCallExpression(importType: ImportType, statement: ASTNode) {
    // making sure we have haven't emitted the same property twice
    if (!statement['emitted']) {
      Object.defineProperty(statement, 'emitted', { enumerable: false, value: true });
      cleanupForTest(statement.arguments[0]);
      cleanupForTest(statement);
      cleanupForTest(statement.callee);

      requireStatementCollection.push({ importType, statement });
    }
  }

  const ast = parseTypeScript(props.code, { jsx: props.jsx });
  const userProps: ISerializableTransformationContext = props.props || {};
  userProps.compilerOptions = props.compilerOptions || {};
  const visitorProps: ITransformerCommon = { onRequireCallExpression, transformationContext: userProps };

  const tranformers = [];
  for (const t of props.transformers) {
    if (t.commonVisitors) {
      tranformers.push(t.commonVisitors(visitorProps));
    }
  }

  transformModule({ root: ast, transformers: tranformers });

  const res = generate(ast, {});

  return { code: res, requireStatementCollection };
}
