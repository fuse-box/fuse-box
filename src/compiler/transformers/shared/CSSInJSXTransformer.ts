import { path2RegexPattern } from '../../../utils/utils';
import { IVisit } from '../../Visitor/Visitor';
import { ASTNode } from '../../interfaces/AST';
import { ITransformer } from '../../interfaces/ITransformer';
import { GlobalContext } from '../../program/GlobalContext';

export interface CSSInJSXTransformerOptions {
  autoInject?: boolean;
  autoLabel?: boolean;
  cssPropOptimization?: boolean;
  emotionCoreAlias?: string;
  jsxFactory?: string;
  labelFormat?: string;
  module?: any;
  sourceMap?: boolean;
  test?: RegExp | string;
}

const labelMapping = {
  '[dirname]': '',
  '[filename]': '',
  '[local]': '',
};

/**
 * 1. compare minifiers styled-components and emotion
 * 2. don't reinvent the wheel..
 * 3. find a place for these helper functions..
 */
// From https://github.com/styled-components/babel-plugin-styled-components/blob/master/src/minify/index.js#L58
// Counts occurences of substr inside str
const countOccurences = (str, substr) => str.split(substr).length - 1;
const compressSymbols = code =>
  code.split(/(\s*[;:{},]\s*)/g).reduce((str, fragment, index) => {
    // Even-indices are non-symbol fragments
    if (index % 2 === 0) {
      return str + fragment;
    }

    // Only manipulate symbols outside of strings
    if (countOccurences(str, "'") % 2 === 0 && countOccurences(str, '"') % 2 === 0) {
      return str + fragment.trim();
    }

    return str + fragment;
  }, '');

// Super simple minifier.. doesn't cover any edge cases or side effects
const minify = (value: string): string => compressSymbols(value.replace(/[\n]\s*/g, ''));

/**
 * @todo
 * 1. expand the minify (way to simple :P)
 *
 * 2. Components as selectors
 * 3. Minification
 * 4. Sourcemaps
 * 5. Minify configurable?
 *
 * 5. Dead Code Elimination // not needed??
 */
export function CSSInJSXTransformer(options?: CSSInJSXTransformerOptions): ITransformer {
  const {
    autoInject = true,
    autoLabel = true,
    // cssPropOptimization = true,
    emotionCoreAlias = '@emotion/core',
    jsxFactory = 'jsx',
    labelFormat = '[dirname]--[local]',
    test = /\.(js|jsx|ts|tsx)$/,
    // sourceMap = true
  } = options;

  const testPathRegex = path2RegexPattern(test);
  const emotionLibraries = [[emotionCoreAlias, 'emotion'], ['@emotion/styled']];
  // const reactLibrary = 'react';

  function renderAutoLabel(): ASTNode {
    return {
      type: 'Literal',
      value: `label:${labelFormat
        .replace(/\[local\]|\[filename\]|\[dirname\]/gi, m => labelMapping[m])
        .replace(/\-\-$/, '')};`,
    };
  }

  return {
    target: { test: testPathRegex },
    commonVisitors: ({ ctx, module }) => {
      // Keep track of imported emotion functions.
      // Allows us to look for custom imports
      // import { css as emotionCss } from '@emotion/core'
      const importedEmotionFunctions = [];
      let needsInjection = true;

      // Check if we're executing an emotion call in this file
      function isEmotionCall(node: ASTNode, index: string): boolean {
        return (
          // css('styles')
          importedEmotionFunctions.indexOf(node[index].name) > -1 ||
          // styled('obj')('styles')
          importedEmotionFunctions.indexOf(node[index].callee && node[index].callee.name) > -1 ||
          // styled.div('div')
          importedEmotionFunctions.indexOf(node[index].object && node[index].object.name) > -1
        );
      }

      const compilerJsxFactory = ctx && ctx.compilerOptions.jsxFactory;

      // Process dirName and fileName only once per file
      const filePath = module.props.fuseBoxPath.replace(/\.([^.]+)$/, '');
      labelMapping['[dirname]'] = filePath.replace(/(\\|\/)/g, '-');
      labelMapping['[filename]'] = filePath.replace(/(.+)(\\|\/)(.+)$/, '$3');

      return {
        onEachNode: (visit: IVisit) => {
          const { node, parent } = visit;

          switch (node.type) {
            case 'CallExpression':
              // append the css class label
              if (autoLabel && !parent.callee && isEmotionCall(node, 'callee')) {
                labelMapping['[local]'] =
                  (parent.type === 'VariableDeclarator' && parent.id.name) ||
                  (parent.type === 'AssignmentExpression' && parent.left.property.name) ||
                  '';
                node.arguments.push(renderAutoLabel());
              }
              break;

            // case 'JSXElement':
            //   if (!cssPropOptimization) {
            //     break;
            //   }
            //   const {
            //     openingElement: { attributes },
            //   } = node;
            //   const attrLength = attributes.length;
            //   if (attrLength === 0) {
            //     break;
            //   }
            //   for (let i = 0; i < attrLength; i++) {
            //     let { name, type, value } = attributes[i]; // call 'attr' once
            //     if (type !== 'JSXAttribute' || name.name !== 'css') {
            //       continue;
            //     }
            //     if (value.expression.type === 'ObjectExpression' || value.expression.type === 'ArrayExpression') {
            //       // css prop optimization
            //       // console.log(name, value);
            //     }
            //   }
            //   break;

            case 'TaggedTemplateExpression':
              if (isEmotionCall(node, 'tag')) {
                let {
                  quasi: { expressions, quasis },
                  tag: callee,
                } = node;

                // Convert template strings to a literal and put the expressions back at it's position
                const styleProperties = [];
                const quasisLength = quasis.length;
                let i = 0;
                while (i < quasisLength) {
                  if (quasis[i].value.cooked) {
                    // We don't need minification in devMode!
                    styleProperties.push({
                      type: 'Literal',
                      value: !ctx.config.isProduction ? quasis[i].value.cooked : minify(quasis[i].value.cooked),
                    });
                  }
                  // Put the expressions back in the place where they belong
                  if (!quasis[i].tail && expressions.length > 0) {
                    styleProperties.push(expressions.shift());
                  }
                  i++;
                }

                // Replace this node with new shiny stuff
                return {
                  replaceWith: {
                    arguments: styleProperties.filter(Boolean),
                    callee,
                    type: 'CallExpression',
                  },
                };
              }
              break;
          }
        },
        onTopLevelTraverse: (visit: IVisit) => {
          const node = visit.node;
          const globalContext = visit.globalContext as GlobalContext;
          if (node.type === 'ImportDeclaration') {
            if (compilerJsxFactory === 'jsx') {
              // @todo:
              // fix this, always insert if it's jsx
            }

            if ([].concat(emotionLibraries[0], emotionLibraries[1]).indexOf(node.source.value) > -1) {
              const specifiersLength = node.specifiers.length;
              let i = 0;
              while (i < specifiersLength) {
                if (node.specifiers[i].imported && node.specifiers[i].imported.name === 'jsx') {
                  needsInjection = false;
                  // @todo:
                  // fix this, remove specifier if compilerJsxFactory === 'jsx'

                  // set the globalContext.jsxFactory for the JSXTransformer
                  globalContext.jsxFactory = node.specifiers[i].local.name;
                  // visit.globalContext.jsxFactory = node.specifiers[i].local.name;
                } else {
                  importedEmotionFunctions.push(node.specifiers[i].local.name);
                }
                i++;
              }

              if (autoInject && needsInjection && emotionLibraries[0].indexOf(node.source.value) > -1) {
                needsInjection = false;

                // set the globalContext.jsxFactory for the JSXTransformer
                if (node.source.value === emotionCoreAlias) {
                  // visit.globalContext.jsxFactory = jsxFactory;
                  globalContext.jsxFactory = jsxFactory;
                }
                node.specifiers.push({
                  imported: {
                    name: 'jsx',
                    type: 'Identifier',
                  },
                  local: {
                    name: jsxFactory,
                    type: 'Identifier',
                  },
                  type: 'ImportSpecifier',
                });
              }
            }
          }
        },
      };
    },
  };
}
