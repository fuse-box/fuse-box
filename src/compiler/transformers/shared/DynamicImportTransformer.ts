import { IVisit, IVisitorMod } from '../../Visitor/Visitor';
import { ASTNode } from '../../interfaces/AST';
import { ITransformer } from '../../interfaces/ITransformer';
import { ImportType } from '../../interfaces/ImportType';
import { getDynamicImport } from '../importHelpers';

export function DynamicImportTransformer(): ITransformer {
  return {
    commonVisitors: props => {
      return {
        onEachNode: (visit: IVisit): IVisitorMod => {
          const node = visit.node;

          const dynamicImport = getDynamicImport(node);
          if (dynamicImport) {
            if (dynamicImport.error) return;

            const requireCallExpression: ASTNode = {
              arguments: [
                {
                  type: 'Literal',
                  value: dynamicImport.source,
                },
              ],
              callee: {
                name: 'require',
                type: 'Identifier',
              },
              type: 'CallExpression',
            };

            if (props.onRequireCallExpression) {
              props.onRequireCallExpression(ImportType.DYNAMIC, requireCallExpression);
            }

            const callExpression: ASTNode = {
              arguments: [
                {
                  async: false,
                  body: requireCallExpression,
                  expression: true,
                  params: [],
                  type: 'ArrowFunctionExpression',
                },
              ],
              callee: {
                computed: false,
                object: {
                  arguments: [],
                  callee: {
                    computed: false,
                    object: {
                      name: 'Promise',
                      type: 'Identifier',
                    },
                    property: {
                      name: 'resolve',
                      type: 'Identifier',
                    },
                    type: 'MemberExpression',
                  },
                  type: 'CallExpression',
                },
                property: {
                  name: 'then',
                  type: 'Identifier',
                },
                type: 'MemberExpression',
              },
              type: 'CallExpression',
            };
            return {
              replaceWith: callExpression,
            };
          }

          return;
        },
      };
    },
  };
}
