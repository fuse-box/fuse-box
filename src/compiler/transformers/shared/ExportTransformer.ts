import { IVisit, IVisitorMod } from '../../Visitor/Visitor';
import {
  ES_MODULE_EXPRESSION,
  createExports,
  createVariableDeclaration,
  isDefinedLocally,
} from '../../Visitor/helpers';
import { ASTNode, ASTType } from '../../interfaces/AST';
import { ITransformer } from '../../interfaces/ITransformer';
import { ImportType } from '../../interfaces/ImportType';
import { GlobalContext } from '../../program/GlobalContext';

const IGNORED_DECLARATIONS = {
  [ASTType.InterfaceDeclaration]: 1,
  [ASTType.TypeAliasDeclaration]: 1,
};
function considerDecorators(node: ASTNode) {
  let statement: ASTNode = node.declaration;
  if (
    node.declaration.type === 'ClassDeclaration' &&
    node.declaration.decorators &&
    node.declaration.decorators.length
  ) {
    statement = createVariableDeclaration(node.declaration.id.name, node.declaration);
    // assigne decorators to the statement
    statement.decorators = node.declaration.decorators;
    statement.$fuse_decoratorForClassIdentifier = node.declaration.id.name;
    // remove them from the original declaration (in order for the decorators to be instered after)
    node.declaration.decorators = [];
  }
  return statement;
}
const INTERESTED_NODES = {
  ExportAllDeclaration: 1,
  ExportDefaultDeclaration: 1,
  ExportNamedDeclaration: 1,
};

export function ExportTransformer(): ITransformer {
  return {
    commonVisitors: props => {
      const definedLocallyProcessed: Record<string, number> = {};
      const compilerOptions = props.compilerOptions;
      let injectEsModuleStatement = compilerOptions.esModuleStatement === true;
      let esModuleStatementInjected = !injectEsModuleStatement;
      return {
        onEachNode: (visit: IVisit) => {
          const global = visit.globalContext as GlobalContext;

          if (global.exportAfterDeclaration) {
            const node = visit.node;
            const definedLocally = isDefinedLocally(node);

            if (definedLocally && !visit.node.$fuse_visited) {
              const newNodes = [];

              const response: IVisitorMod = {};

              for (const localVar of definedLocally) {
                // we don't want to add it multiple times
                if (definedLocallyProcessed[localVar.name]) break;
                definedLocallyProcessed[localVar.name] = 1;
                const targetAfter = global.exportAfterDeclaration[localVar.name];
                if (targetAfter) {
                  if (!response.appendToBody) response.appendToBody = [];
                  for (const item of targetAfter.targets) {
                    const ast = createExports(global.namespace, item, {
                      name: localVar.name,
                      type: 'Identifier',
                    });
                    response.appendToBody.push(ast);
                  }
                }
              }
              if (newNodes.length) {
                node.$fuse_visited = true;

                // console.log('new nodes', newNodes);
                // return { replaceWith: [node].concat(newNodes) };
                response.insertAfterThisNode = newNodes;
              }
              if (response.insertAfterThisNode || response.appendToBody) {
                // make sure exports.__eModule is injected
                if (!esModuleStatementInjected) {
                  esModuleStatementInjected = true;
                  response.prependToBody = [ES_MODULE_EXPRESSION];
                }
                return response;
              }
            }
          }
        },
        onTopLevelTraverse: (visit: IVisit) => {
          const node = visit.node;

          const global = visit.globalContext as GlobalContext;

          const type = node.type;
          if (!INTERESTED_NODES[type]) return;

          // remove export interface
          // export interface HelloWorld{}
          if (type === 'ExportNamedDeclaration') {
            if (node.declaration && IGNORED_DECLARATIONS[node.declaration.type]) {
              return { removeNode: true };
            }
          }

          /**
           * **************************************************************************
           * Export with source
           * export { one, two } from "./source"
           *
           * Needs to be simply replaced, no tracking involved
           */
          if (node.source) {
            const sourceVariable = global.getModuleName(node.source.value);

            // export * from "a"
            if (node.type === 'ExportAllDeclaration') {
              const reqExpression: ASTNode = {
                arguments: [
                  {
                    type: 'Literal',
                    value: node.source.value,
                  },
                ],
                callee: {
                  name: 'require',
                  type: 'Identifier',
                },
                type: 'CallExpression',
              };
              if (props.onRequireCallExpression) {
                props.onRequireCallExpression(ImportType.FROM, reqExpression);
              }
              const response: IVisitorMod = {
                replaceWith: {
                  arguments: [
                    {
                      name: 'exports',
                      type: 'Identifier',
                    },
                    reqExpression,
                  ],
                  callee: {
                    computed: false,
                    object: {
                      name: 'Object',
                      type: 'Identifier',
                    },
                    property: {
                      name: 'assign',
                      type: 'Identifier',
                    },
                    type: 'MemberExpression',
                  },
                  type: 'CallExpression',
                },
              };

              // make sure exports.__eModule is injected
              if (!esModuleStatementInjected) {
                esModuleStatementInjected = true;
                response.prependToBody = [ES_MODULE_EXPRESSION];
              }
              return response;
            }

            // export { foo } from "./bar"
            // creating
            //    var obj = require("module")
            const exportedNodes: Array<ASTNode> = [
              {
                declarations: [
                  {
                    id: {
                      name: sourceVariable,
                      type: 'Identifier',
                    },
                    init: {
                      arguments: [
                        {
                          type: 'Literal',
                          value: node.source.value,
                        },
                      ],
                      callee: {
                        name: 'require',
                        type: 'Identifier',
                      },
                      type: 'CallExpression',
                    },
                    type: 'VariableDeclarator',
                  },
                ],
                kind: 'var',
                type: 'VariableDeclaration',
              },
            ];

            if (type === 'ExportNamedDeclaration') {
              for (const specifier of node.specifiers) {
                exportedNodes.push(
                  createExports(global.namespace, specifier.exported.name, {
                    computed: false,
                    object: {
                      name: sourceVariable,
                      type: 'Identifier',
                    },
                    property: {
                      name: specifier.local.name,
                      type: 'Identifier',
                    },
                    type: 'MemberExpression',
                  }),
                );
              }
            }
            const response: IVisitorMod = { replaceWith: exportedNodes };
            // make sure exports.__eModule is injected
            if (!esModuleStatementInjected) {
              esModuleStatementInjected = true;
              response.prependToBody = [ES_MODULE_EXPRESSION];
            }
            return response;
          }

          /**
           * Handling export default declaratio
           * *********************************************************
           *    export default func
           */
          if (type === 'ExportDefaultDeclaration') {
            if (node.declaration) {
              if (node.declaration.type === ASTType.InterfaceDeclaration) {
                return { removeNode: true };
              }
              // Looking at this case:
              //      console.log(foo)
              //      export default function foo(){}
              if (node.declaration.type === 'FunctionDeclaration' || node.declaration.type === 'ClassDeclaration') {
                if (!node.declaration.id) {
                  node.declaration.id = { name: '__DefaultExport__', type: 'Identifier' };
                }

                // if there are decorators on the class we need to transform and assigne to a variable
                /**
                 * @dec
                 * export default class A {
                 *
                 * }
                 *
                 * Should be converted to:
                 *
                 * let A = class A {
                 * }
                 */
                let statement: ASTNode = considerDecorators(node);
                const response: IVisitorMod = {
                  replaceWith: [
                    statement,
                    createExports(global.namespace, 'default', {
                      name: node.declaration.id.name,
                      type: 'Identifier',
                    }),
                  ],
                };
                // make sure exports.__eModule is injected
                if (!esModuleStatementInjected) {
                  esModuleStatementInjected = true;
                  response.prependToBody = [ES_MODULE_EXPRESSION];
                }
                return response;
              }
              const response: IVisitorMod = {
                replaceWith: createExports(global.namespace, 'default', node.declaration),
              };
              // make sure exports.__eModule is injected
              if (!esModuleStatementInjected) {
                esModuleStatementInjected = true;
                response.prependToBody = [ES_MODULE_EXPRESSION];
              }
              return response;
            }
          }

          if (type === 'ExportNamedDeclaration') {
            /**
             * ******************************************************************************
             * CASE with variable exports
             *
             *      export { name1, name2 }
             *
             * Will be "soft" replaced if those defined earlier
             * Then if it's not defined we need to wait for those (they might come later)
             */

            const newNodes = [];

            if (node.specifiers && node.specifiers.length) {
              for (const specifier of node.specifiers) {
                // if the variable present in the scope...
                // or picked up by the identifierReplacement ( we can leave that for GlobalContextTranformer)
                if (
                  (visit.scope && visit.scope.locals[specifier.local.name]) ||
                  global.identifierReplacement[specifier.local.name]
                ) {
                  newNodes.push(
                    createExports(global.namespace, specifier.exported.name, {
                      name: specifier.local.name,
                      type: 'Identifier',
                    }),
                  );
                } else {
                  // if none was decleared, it must be declared earlier (so we check if later with onEachNode in this transformer )
                  let exportAfterDeclaration = global.exportAfterDeclaration;
                  if (!exportAfterDeclaration) {
                    global.exportAfterDeclaration = exportAfterDeclaration = {};
                  }

                  if (!exportAfterDeclaration[specifier.local.name]) {
                    exportAfterDeclaration[specifier.local.name] = { targets: [] };
                  }
                  // it can have multiple exports of the same variable. For example
                  // export {foo, foo as foo1}

                  exportAfterDeclaration[specifier.local.name].targets.push(specifier.exported.name);
                }
              }
              const response: IVisitorMod = { replaceWith: newNodes };
              // make sure exports.__eModule is injected
              if (!esModuleStatementInjected) {
                esModuleStatementInjected = true;
                response.prependToBody = [ES_MODULE_EXPRESSION];
              }
              return response;
            }
            /**
             * ************************************************************************
             * Export objects directly
             *
             *  export function hello(){}
             *  export class Bar(){}
             */
            if (node.declaration) {
              if (node.declaration.id && node.declaration.id.name) {
                const statement = considerDecorators(node);
                return {
                  replaceWith: [
                    statement,
                    createExports(global.namespace, node.declaration.id.name, {
                      name: node.declaration.id.name,
                      type: 'Identifier',
                    }),
                  ],
                };
              }
              if (node.declaration.type === 'VariableDeclaration') {
                if (node.declaration.declare) {
                  return { ignoreChildren: true, removeNode: true };
                }
                /**
                 * ******************************************************************
                 * Exports defined constants
                 *
                 *    export const foo = 1, bar = 3;
                 */
                if (node.declaration.declarations) {
                  let newNodes = [];
                  for (const declaration of node.declaration.declarations) {
                    if (!declaration.declare) {
                      global.identifierReplacement[declaration.id.name] = {
                        first: global.namespace,
                        second: declaration.id.name,
                      };

                      if (declaration.init) {
                        // check if there is INIT
                        // we night have something like this:
                        //    export var FooBar;
                        newNodes.push(createExports(global.namespace, declaration.id.name, declaration.init));
                      }
                    }
                  }
                  const response: IVisitorMod = { replaceWith: newNodes };

                  // make sure exports.__eModule is injected
                  if (!esModuleStatementInjected) {
                    esModuleStatementInjected = true;
                    response.prependToBody = [ES_MODULE_EXPRESSION];
                  }
                  return response;
                }
              }
            }
          }
        },
      };
    },
  };
}
