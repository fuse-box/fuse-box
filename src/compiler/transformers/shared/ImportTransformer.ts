import * as path from 'path';
import { GlobalContext } from '../../program/GlobalContext';
import { ITransformer } from '../../program/transpileModule';
import { createRequireStatement } from '../../Visitor/helpers';
import { IVisit } from '../../Visitor/Visitor';
import { ITransformerSharedOptions } from '../../interfaces/ITransformerSharedOptions';
import { ImportType } from '../../interfaces/ImportType';

// FIX SHORTHANDS
export function ImportTransformer(options?: ITransformerSharedOptions): ITransformer {
  options = options || {};
  return {
    onTopLevelTraverse: (visit: IVisit) => {
      const node = visit.node;
      const global = visit.globalContext as GlobalContext;

      if (node.type === 'ImportDeclaration') {
        // converts "./foo/bar.hello.js" to foo_bar_hello_js_1 (1:1 like typescript does)
        const variable = path.basename(node.source.value).replace(/\.|-/, '_') + '_' + global.getNextIndex();

        node.specifiers.forEach(specifier => {
          if (specifier.type === 'ImportSpecifier') {
            global.identifierReplacement[specifier.local.name] = {
              first: variable,
              second: specifier.imported.name,
            };
          } else if (specifier.type === 'ImportDefaultSpecifier') {
            global.identifierReplacement[specifier.local.name] = {
              first: variable,
              second: 'default',
            };
          } else if (specifier.type === 'ImportNamespaceSpecifier') {
            // only if we have more than one specifier
            // for instance
            // i=mport MySuperClass, * as everything from "module-name";

            // in every other case like:
            // import * as everything from "module-name";
            // we don't need to do anything, since the variables should match

            global.identifierReplacement[specifier.local.name] = {
              first: variable,
            };

            //if (node.specifiers.length > 1) afterStatement = defineVariable(variable, specifier.local.name);
          }
        });

        return {
          onComplete: () => {
            const reqStatement = createRequireStatement(node.source.value, node.specifiers.length && variable);
            // when everything is finished we need to check if those variables have been used at all
            // they were all unused we need remove the require/import statement at all

            // assuming we have no specififers and this import HAS side effects
            // e.g import "./module"
            const parent = visit.parent;
            const property = visit.property;
            if (node.specifiers.length === 0) {
              const index = parent[property].indexOf(node);
              if (index > -1) {
                if (options.onRequireCallExpression) {
                  options.onRequireCallExpression(ImportType.RAW_IMPORT, reqStatement.reqStatement);
                }
                parent[property].splice(index, 1, reqStatement.statement);
              }
              return;
            }

            let atLeastOneInUse = false;
            for (const specifier of node.specifiers) {
              const traced = global.identifierReplacement[specifier.local.name];
              if (traced && traced.inUse) {
                atLeastOneInUse = true;
                break; // we just need to know if we need to keep the node
              }
            }
            // doing a manual replace
            if (atLeastOneInUse) {
              const index = parent[property].indexOf(node);
              if (index > -1) {
                parent[property].splice(index, 1, reqStatement.statement);
                if (options.onRequireCallExpression) {
                  options.onRequireCallExpression(ImportType.FROM, reqStatement.reqStatement);
                }
              }
            } else {
              const index = parent[property].indexOf(node);
              if (index > -1) parent[property].splice(index, 1);
            }
          },
        };
      }
    },
  };
}
