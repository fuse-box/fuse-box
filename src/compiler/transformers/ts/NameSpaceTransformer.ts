import { ASTNode } from '../../interfaces/AST';
import { ITransformer, transpileModule } from '../../program/transpileModule';
import { IVisit, IVisitorMod } from '../../Visitor/Visitor';
import { GlobalContext, createGlobalContext } from '../../program/GlobalContext';
import { createExports } from '../../Visitor/helpers';

export function NamespaceTransformer(): ITransformer {
  return {
    onEachNode: (visit: IVisit) => {
      if (visit.node.type === 'ModuleDeclaration') {
        return { ignoreChildren: true };
      }
    },
    onTopLevelTraverse: (visit: IVisit): IVisitorMod => {
      let node = visit.node;
      if (node.declare) {
        return { removeNode: true, ignoreChildren: true };
      }
      let withExport = false;
      const globalContext = visit.globalContext as GlobalContext;

      if (node.type === 'ExportNamedDeclaration') {
        if (node.declaration && node.declaration.type === 'ModuleDeclaration') {
          node = node.declaration;
          withExport = true;
        }
      }

      if (node.type === 'ModuleDeclaration') {
        const nm = node.body as ASTNode;
        const mameSpaceName = node.id.name;
        // launch custom transpilation for that namespace
        // we skip children in onEachNode
        transpileModule({
          ...globalContext.programProps,
          ast: nm,
          namespace: mameSpaceName,
          globalContext: createGlobalContext({
            namespace: mameSpaceName,
          }),
        });

        //node.body.context =
        const Declaration: ASTNode = {
          type: 'VariableDeclaration',
          kind: 'var',
          declarations: [
            {
              type: 'VariableDeclarator',
              init: null,
              id: {
                type: 'Identifier',
                name: node.id.name,
              },
            },
          ],
        };

        const FunctionBody: ASTNode = {
          type: 'ExpressionStatement',
          expression: {
            type: 'CallExpression',
            callee: {
              type: 'FunctionExpression',
              params: [
                {
                  type: 'Identifier',
                  name: mameSpaceName,
                },
              ],
              body: {
                type: 'BlockStatement',
                body: (node.body as ASTNode).body,
              },
              async: false,
              generator: false,
              id: null,
            },
            arguments: [
              {
                type: 'LogicalExpression',
                left: {
                  type: 'Identifier',
                  name: mameSpaceName,
                },
                right: {
                  type: 'AssignmentExpression',
                  left: {
                    type: 'Identifier',
                    name: mameSpaceName,
                  },
                  operator: '=',
                  right: {
                    type: 'ObjectExpression',
                    properties: [],
                  },
                },
                operator: '||',
              },
            ],
          },
        };

        const nodes = [Declaration, FunctionBody];
        if (withExport) {
          const exportDeclaration = createExports(globalContext.namespace, mameSpaceName, {
            type: 'Identifier',
            name: mameSpaceName,
          });
          nodes.push(exportDeclaration);
        }
        // replace it with a new node
        return {
          replaceWith: nodes,
        };
      }
    },
  };
}
