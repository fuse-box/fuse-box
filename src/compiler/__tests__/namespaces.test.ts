import { initCommonTransform } from '../testUtils';
import { ExportTransformer } from '../transformers/shared/ExportTransformer';
import { ImportTransformer } from '../transformers/shared/ImportTransformer';
import { NamespaceTransformer } from '../transformers/ts/NameSpaceTransformer';

const testTranspile = (props: { code: string; jsx?: boolean }) => {
  return initCommonTransform({
    jsx: true,
    transformers: [NamespaceTransformer(), ImportTransformer(), ExportTransformer()],
    code: props.code,
  });
};

describe('Namespaces', () => {
  it('should compile namespace', () => {
    const result = testTranspile({
      code: `
        import oi from "./oi";
        export const b = 1;
        namespace Foo {
            export const a = 1;
            export function hello() {}
            console.log(a, oi, b);
        }
        console.log(a, foo, oi, b);
        `,
    });
    expect(result.code).toMatchSnapshot();
  });

  it('should export namespace', () => {
    const result = testTranspile({
      code: `
      export namespace Hey {
        export function hello() {}
      }
        `,
    });
    expect(result.code).toMatchSnapshot();
  });
  // testing all of that is a waste of time, namespaces should be deprecated
});
