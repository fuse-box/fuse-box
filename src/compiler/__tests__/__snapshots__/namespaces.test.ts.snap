// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Namespaces should compile namespace 1`] = `
"var oi_1 = require(\\"./oi\\");
exports.b = 1;
var Foo;
(function (Foo) {
  Foo.a = 1;
  function hello() {}
  Foo.hello = hello;
  console.log(Foo.a, oi_1.default, exports.b);
})(Foo || (Foo = {}));
console.log(a, foo, oi_1.default, exports.b);
"
`;

exports[`Namespaces should export namespace 1`] = `
"var Hey;
(function (Hey) {
  function hello() {}
  Hey.hello = hello;
})(Hey || (Hey = {}));
exports.Hey = Hey;
"
`;
